Switch CLI implemented as per new Design

selvakumarnms opened the pull request:
1. Switch CLI modified to accommodate new Design
2. Switch Extension driver implemented as per new design
3. Mechanism driver modified to get ifindex as part of port provisioning
4. Removed the discovery driver rerences 
5. Added devstack support
6. Removed unwanted DB helper method

manju754 commented:
**Initial Comments**

In baremetal_network_provisioning/common/constants.py
Ln:29,43,44:- changing it to netconf-ssh from netconf_ssh. affects the credential REST API's
why "netconf-ssh" this change is required ?

In baremetal_network_provisioning/db/bm_nw_provision_db.py
Ln:95 switch = query.filter_by(name=name).one() it should be .all()
as 'name' is not unique constraint we can have multiple switches with same name. 

In baremetal_network_provisioning/db/migration/alembic_migrations/versions/3297cd3f2323_bm_nw_provision_models.py
Ln: 42 :- sa.Column('credentials', sa.String(40) make it as 36.

In baremetal_network_provisioning/ml2/extensions/bnp_switch.py
Ln: 126: Check is not required, as it is handled by CLI framework itself.
Ln: 117 and 119:- not required, as we will get only 'id' in the request, we will not get 'name'.
Ln: 206 and 213:- we can have a list of "access_parameters" as we are getting the creds by "name"
Ln:212: if not uuidutils.is_uuid_like(id) it should be 'name' instead of 'id'
Ln: 235: if uuidutils.is_uuid_like(id): everytime we will get 'id' here. the CLI framework resolves 'name' with 'id' and sends the 'id' in REST request

In setup.cfg
Ln:42:- bnp_credential extension is missing

selvakumarnms commented:
updated Manju's comment.

manju754 commented:
In baremetal_network_provisioning/ml2/extensions/bnp_switch.py
Ln:151: . if  'access_parameters' is the list containing more than one creds, we should throw error saying multiple creds exist for the same name and same protocol. As a result we cant go and proceed the switch create. 
we should get only one unique value of cred whether its by 'name' or by 'id' to associate it with switch,
Or it can be handled in "_get_access_param" function also.

Ln:86: "filters = req_dict" should be outside the if condition.

Ln.97 and 101.:- Just a suggestion:
If you return at line 97 inside "if " with list and return at 101 inside "else" without list.
we can avoid Ln 110 and Ln 223

Ln :182 : - do we need to call "_switch_to_show" before returning it to CLI framework?

selvakumarnms commented:
Addressed comments except below:
Ln.97 and 101.:- Just a suggestion:
If you return at line 97 inside "if " with list and return at 101 inside "else" without list.
we can avoid Ln 110 and Ln 223
-- As of now we can keep as it is , as some side effects needs to be tested

Ln :182 : - do we need to call "_switch_to_show" before returning it to CLI framework?
--Not required it is already dict so CLI framework parses the dict in ints formatting.

koteswar16 commented on baremetal_network_provisioning/ml2/extensions/bnp_switch.py:
```diff
@@ -128,288 +105,163 @@ def show(self, request, id, **kwargs):
+            proto_type = access_parameters[0].proto_type
+        else:
+            proto_type = access_parameters.proto_type
         if not access_parameters:
```
remove this check, already doing at L198

koteswar16 commented on baremetal_network_provisioning/ml2/extensions/bnp_switch.py:
```diff
@@ -128,288 +105,163 @@ def show(self, request, id, **kwargs):
+                                                   body['management_protocol'],
+                                                   body['credentials'])
+        credentials = body['credentials']
+        if isinstance(access_parameters, list) and len(access_parameters) > 1:
```
redundant. already did this check inside _get_access_param

koteswar16 commented:
ml2/extensions/bnp_switch.py
L166:get_protocol_validation_result should be changed to get_device_mac
L168: no need of temp var mac_str, error string can be stored in result directly.
L171: initial value of result = SUCCESS
if cond1:
result = ERROR1
else if cond2:
result = ERROR2
............
L172: searching DEVICE_NOT_REACHABLE in value? i think value is last access param right?
L158: why are you storing access params in body?
